### PR TITLE
fix(sysmon): fix MicroPython compilation error when system monitor is enabled

### LINKED ISSUE #5547: Enabled system monitor results compilation error in MicroPython unix build
### LVGL version

v9.0.0

### What happened?

When system monitor is enabled in MicroPython binding, then the unix port build fails with following error:
```
CC build-standard/lvgl/lv_mpy.c
build-standard/lvgl/lv_mpy.c:21405:22: error: ‘lv_sysmon_class’ undeclared here (not in a function); did you mean ‘lv_sysmon_create’?
21405 |     .lv_obj_class = &lv_sysmon_class,
      |                      ^~~~~~~~~~~~~~~
      |                      lv_sysmon_create
-e See https://github.com/micropython/micropython/wiki/Build-Troubleshooting
make: *** [../../py/mkrules.mk:83: build-standard/build-standard/lvgl/lv_mpy.o] Error 1
```

There was a refactor of system monitor, and since then there is no `lv_sysmon_class`: https://github.com/lvgl/lvgl/commit/a61d87f8403d7be17f238640b520ca0fd96c5c57#diff-931dc821c9118f8f5432800178c30f1c9bc562ad1c72ff51ebdd8c0aed5232d9L61

### How to reproduce?

1. Clone [lv_micropython](https://github.com/lvgl/lv_micropython) repository (`master` branch)

2. Enable system monitor in `lv_conf.h` file: 
```
#define LV_USE_SYSMON 1
#define LV_USE_PERF_MONITOR 1
```

3. Build unix port in `lv_micropython` repo directory:
```
make -C mpy-cross
cd ports/unix
make submodules
make
```

### PR DESCRIPTION
**Pull request by PGNetHun**
### Description of the feature or fix

Fix for https://github.com/lvgl/lvgl/issues/5547

When system monitor is enabled in MicroPython binding, then the unix port build fails with following error:
```
CC build-standard/lvgl/lv_mpy.c
build-standard/lvgl/lv_mpy.c:21405:22: error: ‘lv_sysmon_class’ undeclared here (not in a function); did you mean ‘lv_sysmon_create’?
21405 |     .lv_obj_class = &lv_sysmon_class,
      |                      ^~~~~~~~~~~~~~~
      |                      lv_sysmon_create
-e See https://github.com/micropython/micropython/wiki/Build-Troubleshooting
make: *** [../../py/mkrules.mk:83: build-standard/build-standard/lvgl/lv_mpy.o] Error 1
```

There was a refactor of system monitor, and since then there is no `lv_sysmon_class`: https://github.com/lvgl/lvgl/commit/a61d87f8403d7be17f238640b520ca0fd96c5c57#diff-931dc821c9118f8f5432800178c30f1c9bc562ad1c72ff51ebdd8c0aed5232d9L61

Build of other ports fail, too.

### Notes
- Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed.
- Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant.
- Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/master/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/master/Kconfig).
- Run `scripts/code-format.py` ([astyle](http://astyle.sourceforge.net/install.html) version [v3.4.10](https://github.com/szepeviktor/astyle/releases/tag/v3.4.10) needs to be installed) and follow the [Code Conventions](https://docs.lvgl.io/master/CODING_STYLE.html).
- Mark the Pull request as [Draft](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request) while you are working on the first version, and mark is as _Ready_ when it's ready for review.
- When changes were requested, [re-request review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) to notify the maintainers.
- Help us to review this Pull Request! Anyone can [approve or request changes](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews).

### CODE REVIEW

**Review by C47D — APPROVED** [review]

File: src/others/sysmon/lv_sysmon.c [hunk 1]
```diff
@@ -163,7 +167,7 @@ static void perf_update_timer_cb(lv_timer_t * t)
     if(!sysmon_perf.inited && lv_display_get_default()) {
         lv_display_add_event_cb(lv_display_get_default(), perf_monitor_disp_event_cb, LV_EVENT_ALL, NULL);
 
-        lv_obj_t * obj1 = lv_sysmon_create(lv_layer_sys());
```
Comment (kisvegabor) [inline src/others/sysmon/lv_sysmon.c:166]: I vote for renaming to `lv_sysmon_show` to not follow the widget-like API for non-widgets.

What do you think?

Comment (kdschlosser) [reply src/others/sysmon/lv_sysmon.c:166]: works for me. The sysmon is not a widget so it really shouldn't use the same kind of naming convention as a widget.

Comment (PGNetHun) [reply src/others/sysmon/lv_sysmon.c:166]: OK, I will add show and hide functions (for both performance and memory) in the next days.

**Comment by PGNetHun** [comment]
Actually the change was simpler than I thought, so these public functions are added:
- lv_sysmon_show_performance()
- lv_sysmon_hide_performance()
- lv_sysmon_show_memory()
- lv_sysmon_hide_memory()

If sysmon is enabled (either for performance or memory), then show functions will be called automatically -> this grants the sysmon to be "backward compatible"

Tested with MicroPython, and it works:
image


Some remarks:
- in MicroPython sysmon memory is 0, so practically useless
- FPS calculation is a bit strange: I assume it calculates the LVGL rendering time, not the real display refresh rate: just add a 240x240 canvas, draw on it, and you will see, that LVGL rendering time will be low, and FPS high, but yet on ESP32 LCD the refresh rate feels much lower (< 20 fps)

File: src/others/sysmon/lv_sysmon.c (hunk 1 above)
Comment (PGNetHun) [reply src/others/sysmon/lv_sysmon.c:166]: Functions are added

**Comment by kdschlosser** [comment]
> in MicroPython sysmon memory is 0, so practically useless

That's because in MicroPython the memory allocation isn't handled by LVGL. It is handled using the stdlib.

If the code in LVGL had a define set up that pointed to the function that returns the memory use for the sysmon we could write a function that would collect the memory stats from micropython so the sysmon memory would work properly.

**Comment by kdschlosser** [comment]
LVGL's framerate is going to be skewed. It is not calculating for a total refresh of the display. the binding doesn't do full display redraws. it uses a fraction of the display size. so a frame is counted when only one of these small frame buffers are written.

I am not sure if the framerate portion of the sysmon was ever fixed because it would count calls to a function as a frame written even if there as no data to write. That would also skew the number as well. The 3rd thing is it is assumed that task_handler is always being called at the exact time the timer would expire to update the framerate. That number was hard coded into the calculations so that would also skew it as well.

I had opened an issue about it some time ago, I don't recall if it ever got corrected.

File: src/others/sysmon/lv_sysmon.h [hunk 2]
```diff
@@ -77,20 +78,37 @@ typedef struct {
...
+
+#if LV_USE_MEM_MONITOR
 
 /**
- * Create a system monitor object.
- * @param parent pointer to an object, it will be the parent of the new system monitor
- * @return       pointer to the new system monitor object
+ * Show system memory monitor: used memory and the memory fragmentation
  */
-lv_obj_t * lv_sysmon_create(lv_obj_t * parent);
```
Comment (FASTSHIFT) [inline src/others/sysmon/lv_sysmon.h:86]: It is recommended to retain `lv_sysmon_create` so that users can implement their own monitors.

**Review by FASTSHIFT — CHANGES_REQUESTED** [review]

**Comment by kisvegabor** [comment]
@kdschlosser 
> LVGL's framerate is going to be skewed. 

It was fixed in v9 and it's confirmed from many sources that it's quite accurate now.

File: src/others/sysmon/lv_sysmon.c [hunk 3]
```diff
@@ -93,7 +96,65 @@ void _lv_sysmon_builtin_deinit(void)
 #endif
 }
 
-lv_obj_t * lv_sysmon_create(lv_obj_t * parent)
+#if _USE_PERF_MONITOR
+
+void lv_sysmon_show_performance(void)
```
Comment (kisvegabor) [inline src/others/sysmon/lv_sysmon.c:101]: `create` works here as this API cannot be interpreted as widget create API.
```suggestion
void lv_sysmon_create_performance(void)
```

File: src/others/sysmon/lv_sysmon.h (hunk 2 above)
Comment (kisvegabor) [reply src/others/sysmon/lv_sysmon.h:86]: Is it a real use case?

**Comment by kdschlosser** [comment]
> It was fixed in v9 and it's confirmed from many sources that it's quite accurate now.

OK cool I wasn't sure if it had been fixed or not. I am not sure why @PGNetHun says it is not reading correctly.

File: src/others/sysmon/lv_sysmon.h (hunk 2 above)
Comment (FASTSHIFT) [reply src/others/sysmon/lv_sysmon.h:86]: Yes, it is necessary if we have multiple heaps or multiple resources that need to be dynamically monitored.

Comment (kisvegabor) [reply src/others/sysmon/lv_sysmon.h:86]: In this case we might be better converting it widget-like by adding a class definition.

**Comment by kisvegabor** [comment]
@PGNetHun are you planing to finalize this update? We can take it over if needed.

**Comment by PGNetHun** [comment]
> @PGNetHun are you planing to finalize this update? We can take it over if needed.

Hello, yes, sorry.

**Comment by PGNetHun** [comment]
> > @PGNetHun are you planing to finalize this update? We can take it over if needed.
> 
> Hello, yes, sorry, I will continue

Sorry @kisvegabor ,
but I can't (and don't have time) to change sysmon to widget-like control.
Please continue the code changes.

**Comment by kisvegabor** [comment]
Okay, I've opened #6073.

**Pull request closed without merging**